File: src/java/voldemort/store/routed/RoutedStore.java
/*
 * Copyright 2008-2010 LinkedIn, Inc
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package voldemort.store.routed;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import org.apache.log4j.Logger;

import voldemort.VoldemortException;
import voldemort.client.TimeoutConfig;
import voldemort.cluster.Cluster;
import voldemort.cluster.failuredetector.FailureDetector;
import voldemort.routing.RoutingStrategy;
import voldemort.routing.RoutingStrategyFactory;
import voldemort.store.AbstractStore;
import voldemort.store.NoSuchCapabilityException;
import voldemort.store.Store;
import voldemort.store.StoreCapabilityType;
import voldemort.store.StoreDefinition;
import voldemort.utils.ByteArray;
import voldemort.utils.Time;
import voldemort.utils.Utils;

/**
 * A Store which multiplexes requests to different internal Stores
 * 
 * 
 */
public abstract class RoutedStore extends AbstractStore<ByteArray, byte[], byte[]> {

    protected final Map<Integer, Store<ByteArray, byte[], byte[]>> innerStores;
    protected final boolean repairReads;
    protected final ReadRepairer<ByteArray, byte[]> readRepairer;
    protected final TimeoutConfig timeoutConfig;
    protected final Time time;
    protected final StoreDefinition storeDef;
    protected final FailureDetector failureDetector;
    protected volatile RoutingStrategy routingStrategy;
    protected final Logger logger = Logger.getLogger(getClass());

    protected RoutedStore(String name,
                          Map<Integer, Store<ByteArray, byte[], byte[]>> innerStores,
                          Cluster cluster,
                          StoreDefinition storeDef,
                          boolean repairReads,
                          TimeoutConfig timeoutConfig,
                          FailureDetector failureDetector,
                          Time time) {
        super(name);
        if(storeDef.getRequiredReads() < 1)
            throw new IllegalArgumentException("Cannot have a storeDef.getRequiredReads() number less than 1.");
        if(storeDef.getRequiredWrites() < 1)
            throw new IllegalArgumentException("Cannot have a storeDef.getRequiredWrites() number less than 1.");
        if(storeDef.getPreferredReads() < storeDef.getRequiredReads())
            throw new IllegalArgumentException("storeDef.getPreferredReads() must be greater or equal to storeDef.getRequiredReads().");
        if(storeDef.getPreferredWrites() < storeDef.getRequiredWrites())
            throw new IllegalArgumentException("storeDef.getPreferredWrites() must be greater or equal to storeDef.getRequiredWrites().");
        if(storeDef.getPreferredReads() > innerStores.size())
            throw new IllegalArgumentException("storeDef.getPreferredReads() is larger than the total number of nodes!");
        if(storeDef.getPreferredWrites() > innerStores.size())
            throw new IllegalArgumentException("storeDef.getPreferredWrites() is larger than the total number of nodes!");

        this.innerStores = new ConcurrentHashMap<Integer, Store<ByteArray, byte[], byte[]>>(innerStores);
        this.repairReads = repairReads;
        this.readRepairer = new ReadRepairer<ByteArray, byte[]>();
        this.timeoutConfig = timeoutConfig;
        this.time = Utils.notNull(time);
        this.storeDef = storeDef;
        this.failureDetector = failureDetector;
        this.routingStrategy = new RoutingStrategyFactory().updateRoutingStrategy(storeDef, cluster);
    }

    public void updateRoutingStrategy(RoutingStrategy routingStrategy) {
        logger.info("Updating routing strategy for RoutedStore:" + getName());
        this.routingStrategy = routingStrategy;
    }

    @Override
    public void close() {
        VoldemortException exception = null;

        for(Store<?, ?, ?> store: innerStores.values()) {
            try {
                store.close();
            } catch(VoldemortException e) {
                exception = e;
            }
        }

        if(exception != null)
            throw exception;
    }

    public Map<Integer, Store<ByteArray, byte[], byte[]>> getInnerStores() {
        return this.innerStores;
    }

    @Override
    public Object getCapability(StoreCapabilityType capability) {
        switch(capability) {
            case ROUTING_STRATEGY:
                return this.routingStrategy;
            case READ_REPAIRER:
                return this.readRepairer;
            case VERSION_INCREMENTING:
                return true;
            default:
                throw new NoSuchCapabilityException(capability, getName());
        }
    }

}

File: bindings/cpp/src/RoutedStore.cpp
/* -*- C++ -*-; c-basic-offset: 4; indent-tabs-mode: nil */
/*
 * Implementation for RoutedStore class.
 * 
 * Copyright (c) 2009 Webroot Software, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

#include "RoutedStore.h"
#include "voldemort/UnreachableStoreException.h"
#include "voldemort/InsufficientOperationalNodesException.h"
#include <iostream>

namespace Voldemort {

using namespace boost;
using namespace std;

static const bool REPAIR_READS = true;

RoutedStore::RoutedStore(const std::string& storeName,
                         shared_ptr<ClientConfig>& config,
                         shared_ptr<Cluster>& clust,
                         shared_ptr<std::map<int, shared_ptr<Store> > >& map,
                         shared_ptr<RoutingStrategy>& routingStrat)
    : name(storeName), clientConfig(config), cluster(clust), clusterMap(map), 
      routingStrategy(routingStrat) {

}

RoutedStore::~RoutedStore() {
    close();
}

static bool doGetFromStore(const std::string& key, 
                           std::list<VersionedValue>** result,
                           Node* node,
                           Store* store) {
    *result = NULL;
    try {
        *result = store->get(key);
        node->setAvailable(true);
        return true;
    } catch (UnreachableStoreException& e) {
        /* XXX - TODO add real logging */
        std::cerr << "WARNING: Could not read node: " << e.what() << std::endl;
        node->setAvailable(false);
    }
    return false;
}

std::list<VersionedValue>* RoutedStore::get(const std::string& key) {
    std::list<VersionedValue>* result = NULL;
    bool status = false;
    {
        /* Start by routing to the preferred list one at a time. */
        RoutingStrategy::prefListp prefList = routingStrategy->routeRequest(key);
        RoutingStrategy::prefList::const_iterator it;
        for (it = prefList->begin(); it != prefList->end(); ++it) {
            status = doGetFromStore(key, &result,
                                    it->get(),
                                    (*clusterMap)[(*it)->getId()].get());
            if (status) return result;
        }
    }
    {
        /* If that fails just try every node in the cluster */
        const Cluster::nodeMap* nm = cluster->getNodeMap();
        Cluster::nodeMap::const_iterator it;
        for (it = nm->begin(); it != nm->end(); ++it) {
            if (it->second.get()->isAvailable(clientConfig->getNodeBannageMs())) {
                status = doGetFromStore(key, &result,
                                        it->second.get(),
                                        (*clusterMap)[it->first].get());
            }
            if (status) return result;
        }
    }
    
    throw InsufficientOperationalNodesException("Could not reach any "
                                                "node for get operation");
}

static bool doPutFromStore(const std::string& key, 
                           const VersionedValue& value,
                           Node* node,
                           Store* store) {
    try {
        store->put(key, value);
        node->setAvailable(true);
        return true;
    } catch (UnreachableStoreException& e) {
        node->setAvailable(false);
    }
    return false;
}

void RoutedStore::put(const std::string& key, const VersionedValue& value) {
    bool status = false;
    {
        /* Start by routing to the preferred list one at a time. */
        RoutingStrategy::prefListp prefList = routingStrategy->routeRequest(key);
        RoutingStrategy::prefList::const_iterator it;
        for (it = prefList->begin(); it != prefList->end(); ++it) {
            status = doPutFromStore(key, value,
                                    it->get(),
                                    (*clusterMap)[(*it)->getId()].get());
            if (status) return;
        }
    }
    {
        /* If that fails just try every node in the cluster */
        const Cluster::nodeMap* nm = cluster->getNodeMap();
        Cluster::nodeMap::const_iterator it;
        for (it = nm->begin(); it != nm->end(); ++it) {
            if (it->second.get()->isAvailable(clientConfig->getNodeBannageMs())) {
                status = doPutFromStore(key, value,
                                        it->second.get(),
                                        (*clusterMap)[it->first].get());
            }
            if (status) return;
        }
    }
    
    throw InsufficientOperationalNodesException("Could not reach any "
                                                "node for put operation");
}

static bool doDeleteFromStore(const std::string& key, 
                              const Version& version,
                              bool* result,
                              Node* node,
                              Store* store) {
    try {
        *result = store->deleteKey(key, version);
        node->setAvailable(true);
        return true;
    } catch (UnreachableStoreException& e) {
        node->setAvailable(false);
    }
    return false;
}

bool RoutedStore::deleteKey(const std::string& key, const Version& version) {
    bool status = false;
    bool result = false;
    {
        /* Start by routing to the preferred list one at a time. */
        RoutingStrategy::prefListp prefList = routingStrategy->routeRequest(key);
        RoutingStrategy::prefList::const_iterator it;
        for (it = prefList->begin(); it != prefList->end(); ++it) {
            status = doDeleteFromStore(key, version, &result,
                                       it->get(),
                                       (*clusterMap)[(*it)->getId()].get());
            if (status) return result;
        }
    }
    {
        /* If that fails just try every node in the cluster */
        const Cluster::nodeMap* nm = cluster->getNodeMap();
        Cluster::nodeMap::const_iterator it;
        for (it = nm->begin(); it != nm->end(); ++it) {
            if (it->second.get()->isAvailable(clientConfig->getNodeBannageMs())) {
                status = doDeleteFromStore(key, version, &result,
                                           it->second.get(),
                                           (*clusterMap)[it->first].get());
            }
            if (status) return result;
        }
    }
    
    throw InsufficientOperationalNodesException("Could not reach any "
                                                "node for delete operation");
}

const std::string* RoutedStore::getName() {
    return &name;
}

void RoutedStore::close() {
    std::map<int, shared_ptr<Store> >::const_iterator it;
    for (it = clusterMap->begin(); it != clusterMap->end(); ++it) {
        it->second->close();
    }
}


} /* namespace Voldemort */
